Dashboard refresh functionality

CMDRyan opened the pull request:
@Steven-Chan 

- Server side API has been created
- The browser will now call API and get data every 10 seconds
   - A function "_autoupdate()_" is created and will be called on load to start the reload cycle 
   - Numbers that indicate jobs.length will now be called by function instead.

Minor fixes:

- Removed unnecessary lines from application
- Added details to error page

*This pull request is based on pull request: "_Implement cache data to improve app.js efficiency #35_"

Steven-Chan commented on src/views/pages/index.ejs:
```diff
@@ -21,26 +20,45 @@
+        document.getElementById(columnName).innerHTML = JSON.parse(this.response).join("");
+        document.getElementById(`${columnName}-length`).innerHTML = JSON.parse(this.response).length;
+      }
+      xhttp.open("GET", "<%= origin %>/api/groups/794774?status=" + columnName);
```
The url should be constructed with the browser `URL` instead of string concat.

CMDRyan commented:
@Steven-Chan 

app.js
- Reduce cache timeout to 5 seconds

Index.ejs:
- Refresh time will now align with cache
- Removed reloadCol call on autoUpdate load
- JSON.parse will only be called once
- API url and group id are now constructed with `window.location`
- Original cards are now recovered

singleJobCard.ejs:
- Renamed "status" to "job"

Needs to be done:
Check mouse idle(e.g. 30 seconds), pause refresh and resume on mousemove

Steven-Chan commented on src/app.js:
```diff
@@ -27,6 +30,15 @@ app.set("view engine", "ejs");
+});
+const getJobs = withCache(getData.getJobs, {
+  cacheStorage: jobsCache,
+  ttl: 5,
```
Now this is completely detached from the default auto reload interval. They should share the same config, so that developers would not miss when updating the value.

BTW, ideally both of the cache ttl should come from server config.

Steven-Chan replied:
@CMDRyan This one is not resolved yet btw.

CMDRyan commented:
@Steven-Chan 

- Index will now check if user has gone idle every 30 seconds
- ...and will stop refreshing 
- Index will resume refreshing by checking mouse movement

Steven-Chan reviewed (changes requested):
Note for change request talked offline.

1. Always use URL instead of string to construct and manipulate the URL.
2. Refactor the client side js code to unobtrusive style. I expect there should be two set of logic needs to be refactored. 1.) ajax API call to update html and 2.) user inactivity detection.

CMDRyan commented:
@Steven-Chan 

- Timeout and refresh time are now aligned, and are required to be defined in .env
- API call url is now constructed with URL object

Steven-Chan commented on src/views/pages/index.ejs:
```diff
@@ -21,26 +20,74 @@
     </footer>
+  </div>
+  <script>
+    function buildApiUrl(columnName){
+      const apiUrl = new URL(`/api`+document.location.pathname, document.location.origin)
+      apiUrl.searchParams.append(`status`,columnName)
+      document.getElementById(columnName).setAttribute("data-ajax-url", apiUrl);
+    }
+  </script>
+   <script>
+    var userIsActive;
+    var timeout;
+    buildApiUrl("created")
+    buildApiUrl("pending")
+    buildApiUrl("running")
```
Don't use JS to build this. If you really want a clean DRY, you generate it with html template.

However, I would recommend just explicitly hard code the url on html. Because, from my experience, it is not wise to write very DRY code on UI. UI depends on user requirement and user requirement changes a lot. DRY code is built on top of assumption, and if the assumption breaks, you will need to refactor.

Steven-Chan commented on src/views/pages/index.ejs:
```diff
@@ -21,26 +20,74 @@
+    buildApiUrl("created")
+    buildApiUrl("pending")
+    buildApiUrl("running")
+    $.fn.extend({
```
Why define the autoUpdate on jQuery when this is not about jQuery?

Steven-Chan commented on src/views/pages/index.ejs:
```diff
@@ -21,26 +20,74 @@
+    buildApiUrl("running")
+    $.fn.extend({
+
+      autoUpdate: function(){
```
`autoUpdate` is not a good function name, it is doing something, but it is not a verb. Usually a function with a noun as function name is for attribute getter.

Steven-Chan commented on src/views/pages/index.ejs:
```diff
@@ -21,26 +20,74 @@
+          }, <%= cache_timeout %>);
+      }(),
+    
+      userActive: function(){
```
Similar to `autoUpdate`, this is not a good function name.

CMDRyan commented:
@Steven-Chan 

- Renamed functions
- autoUpdate() no longer extends jQuery
- Removed unused unobtrusive library and files
- Reword pathname of api
- Refactor api url in index

CMDRyan commented:
@Steven-Chan 

- Fixed card's link not working issue

CMDRyan commented:
@Steven-Chan 

- Further hardcoded api URLs

CMDRyan commented:
@Steven-Chan 

- Added handler in library with additional tag condition for preventDefault click issue  
- Created userActivity object with set and get activity method
- isUserActive is no longer a global variable
- Index will now call userActivity to reset timeout and get activity
- Reverted api pathname
